### ISSUE TITLE
Multi-TFM frameworks stop restoring

### ISSUE THREAD

**Issue by rohit21agrawal** [dotnet/project-system#1503]
_From @jinujoseph on February 2, 2017 21:14_

- Create a .Net standard class library 
- Edit the csproj and change TFM to 
`<TargetFrameworks>netstandard1.4</TargetFrameworks>`
will be promoted to reload 
Build 
- Edit the csproj and change to 
` <TargetFrameworks>netstandard1.4;netstandard1.0</TargetFrameworks>`
Build 

**Expected** 
Builds successfully 

**Actual** 

![image](https://cloud.githubusercontent.com/assets/12449387/22568433/b2760ecc-e948-11e6-9476-7e558af8866e.png)

> 1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(76,5): error : Assets file 'c:\users\jinuj\documents\visual studio 2017\Projects\ClassLibrary10\ClassLibrary10\obj\project.assets.json' doesn't have a target for '.NETStandard,Version=v1.0'. Ensure you have restored this project for TargetFramework='netstandard1.0' and RuntimeIdentifier=''.

I have tried restoring the solution and rebuilding , which doesn't resolve the issue
**Note**: Closing and reopening the project fixes it.

**Build**: 26201.01.d15rel

_Copied from original issue: NuGet/Home#4479_

**Comment by rohit21agrawal** [comment]
_From @davkean on February 3, 2017 0:10_

Changing the TFM -> TFMs, should have prompted a reload, no?

**Comment by rohit21agrawal** [comment]
Upon investigation it turns out that the project isn't getting nominated in the following scenarios: 

1) When TargetFrameworks is set to netstandard1.4 , and I add another framework say, netstandard1.0
2) adding a property like Version doesn't nominate the project.

**Comment by srivatsn** [comment]
@mavasani is this fixed with your PR to clear the contexts?

**Comment by mavasani** [comment]
@Srivatsn, no my change is specific to editing "TargetFramework" not "TargetFrameworks"

**Comment by srivatsn** [comment]
@mavasani Should you be including changes to TargetFrameworks as well?

@rohit21agrawal  - Looking at NuGetRestore.xaml[https://github.com/dotnet/roslyn-project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/NuGetRestore.xaml] Version is one of the properties for which we nominate - so i'm surprised. @natidea can you look into that?

**Comment by mavasani** [comment]
@srivatsn - the scenario here is specific to build, and my changes are specific to design time build in LanguageService. Design time builds work fine for TargetFrameworks edit case as each framework dimension creates a different configured project, which is not the case when TargetFramework is edited.

**Comment by srivatsn** [comment]
That's right - my mistake but I think it's related. My guess is that there's either a bad design time build or a bad update and we're nominating restore. Closing and reopening the project works because we nominate restore.

**Comment by natidea** [comment]
This seems to have regressed. When I modify TargetFrameworks, the subscriptions are reset as expected -- however a subscription update is not being fired immediately after the reset, as it always has before. Subsequent changes also are not firing which indicates the new subscriptions are not taking.

**Comment by davkean** [comment]
That's probably a good sign that the block has failed probably due an unhandled exception./

**Comment by natidea** [comment]
I have not been able to detect any exceptions in the debugger around this scenario.

I notice that if I start out with one TFM pair, (e.g. `<TargetFrameworks>netstandard1.4;netstandard1.0</TargetFrameworks>`), changing it to a different pair
(e.g. `<TargetFrameworks>netstandard1.4;netstandard1.3</TargetFrameworks>`) does not trigger a subscription update. But changing it back to `netstandard1.4;netstandard1.0` does. This is despite the fact that we are disposing these subscriptions each time we reset. Its like something deeper in the stack is holding on to the old subscription.

**Comment by davkean** [comment]
Talking to @lifengl, this is because we're [expecting](https://github.com/dotnet/roslyn-project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs#L96-L122) configurations to agree on their _configured project version_. These, however, are you unique versions per configured project and aren't expected to match. It's a coincidence that they do match in these cases where the event is fired.

Here's our offline conversation:

```
[‎02/‎14/‎17 12:15 PM] Lifeng Lu: 
the problem is an old design problem in our code
we predefined some data source version keys, like the configuredProjectVersion
[‎02/‎14/‎17 12:16 PM] Lifeng Lu: 
all configured project uses the same key for that... it is hard to change that, because it is a part of the public API
many code uses it
when you merge data from those configure projects, we don't know which one is which
[‎02/‎14/‎17 12:18 PM] Lifeng Lu: 
syncLinkTo will require all of them match before publishing data to the next level, because it wants data from all sources consistent (basically, it must have the same version number for all data source version keys)
but each configured project maintains its own configuredProjectVersion, they may never match
so your dataflow will be blocked forever
[‎02/‎14/‎17 12:19 PM] 
Yeah, that's what we're seeing
[‎02/‎14/‎17 12:19 PM] 
we want them to agree on the underlying project file version
[‎02/‎14/‎17 12:20 PM] Lifeng Lu: 
we thought about that (I thought we discussed with Anton), and think we eventually want to use a different key for different configured project
for all independent version number
but maybe still need retain old public version number there
[‎02/‎14/‎17 12:21 PM] Lifeng Lu: 
to make your code to work, you must do some workaround
[‎02/‎14/‎17 12:22 PM] Lifeng Lu: 
either you don't use the syncLinkTo at all
I mean, your code will handle changes from all projects, and handle them directly
anyway, how much benefit you get from syncLinkTo?
[‎02/‎14/‎17 12:24 PM] 
We just want to be called back once when they all agree on data
[‎02/‎14/‎17 12:25 PM] Lifeng Lu: 
Hmm... I don't think we have a way to know how to align different configurations after one change :(
seems to be a gap in the design
[‎02/‎14/‎17 12:27 PM] Lifeng Lu: 
but maybe your code can wait the dataflow carries the configuredProjectVersion matching the current value of the configured project, and skip all the middle state?
[‎02/‎14/‎17 12:29 PM] Lifeng Lu: 
if you want to use syncLinkTo, you can add an extra block to remove all conflicting versions (maybe mapping them to different keys) before feeding a syncLinkTo block,
so it can merge unconfigured project level versions
[‎02/‎14/‎17 12:30 PM] 
But what exactly would they sync on?
[‎02/‎14/‎17 12:35 PM] Lifeng Lu: 
some unconfigured project level data sources, like the MSBUild GlobalProject properties
[‎02/‎14/‎17 12:37 PM] 
There's not something that represents the project file version?
[‎02/‎14/‎17 12:38 PM] 
Or is this a concept that only occurs at the configured project level
[‎02/‎14/‎17 12:38 PM] 
?
[‎02/‎14/‎17 12:39 PM] Lifeng Lu: 
it only occurs at the configured project level, no file version

```

**Comment by srivatsn** [comment]
We need to sit down and design the fix here and given the workaround, that's not going to meet the bar for RTM. Moving to 1.1.

**Comment by davkean** [comment]
Looking through CPS, I found the following code which is an interesting way to represent a "project version" (vs a "config version"):

``` C#
        /// <summary>
        /// Responds to direct project file changes by incrementing an opaque version number.
        /// </summary>
        private Task ProjectLockService_ReleasingWriteLockAsync(object sender, ReleasingWriteLockEventArgs e)
        {
            if (e.WrittenProjectXmlPaths.Contains(this.UnconfiguredProject.FullPath))
            {
                Interlocked.Increment(ref this.projectVersion);
            }

            return TplExtensions.CompletedTask;
        }
```

**Comment by srivatsn** [comment]
The scenario has been fixed with #2298. However https://github.com/dotnet/project-system/issues/2324 tracks making a more complete fix for it.
